envoy/filter/config_provider_manager.h

```diff
@@ -23,7 +23,7 @@ using DynamicFilterConfigProviderPtr = std::unique_ptr<DynamicFilterConfigProvid
  * The FilterConfigProviderManager exposes the ability to get an FilterConfigProvider
  * for both static and dynamic filter config providers.
  */
-template <class FactoryCb> class FilterConfigProviderManager {
+template <class FactoryCb, class FactoryCt> class FilterConfigProviderManager {
```

Comment:
This looks suspicious. ECDS subscriptions should outlive parent objects because they may be shared. If you use any other context other than Server then the parent object may become dangling.

E.g. if L1 and L2 share a subscription, L1 creating with its listener context, then L1 removed but L2 continuing to use the  subscription. The subscription should not capture anything from L1.

Reply:
Thanks for the comments.

This change didn't modify the behavior of user ServerFactoryContext in subscription.  It just abstracted the way how to create the parent object (filter_config_provider_manager_), i.e by passing in either Server::Configuration::FactoryContext for HTTP filter or Server::Configuration::ListenerFactoryContext for listener filter.  Underline, they still derive the ServerFactoryContext from the passed in  factory context by factory_context.getServerFactoryContext() to create subscriptions:  https://github.com/envoyproxy/envoy/blob/4dc78ded31fb64f5790553bd67e4f3bc9d13998e/source/common/filter/config_discovery_impl.cc#L190. 

In your example, if L1 and L2 share a subscription, even L1 and L2 have different FactoryContext, as long as they have same ServerFactoryContext, L1 removal won't impact L2 continue use the subscription since ServerFactoryContext will stay https://github.com/envoyproxy/envoy/blob/4dc78ded31fb64f5790553bd67e4f3bc9d13998e/envoy/server/factory_context.h#L149.

In a nutshell, this change is to abstract the FactoryContext out to make class FilterConfigProviderManager and some corresponding classes be able to support both HTTP and listener filter.  There should be no behavior changes for subscription or how HTTP filter works.

Reply:
I see, thanks for explaining. So this is not a problem then. Would it be cleaner to just pass `ServerFactoryContext` as the argument type and modify the callers?

Reply:
That might not be doable since createDynamicFilterConfigProvider() method took in FactoryContext, not ServerFactoryContext as 
https://github.com/envoyproxy/envoy/blob/4dc78ded31fb64f5790553bd67e4f3bc9d13998e/source/extensions/filters/network/http_connection_manager/config.cc#L685

And the FactoryContext parameter is used in a lot of other places in this method not related to subscription, e.g.:

https://github.com/envoyproxy/envoy/blob/4dc78ded31fb64f5790553bd67e4f3bc9d13998e/source/common/filter/config_discovery_impl.h#L335

Reply:
Ok, that's fine. Thanks for investigating.

Reply:
The warming aspect specifically does require parent context, so this templating is in fact needed.